[BE][DTensor] fix DTensor equal op

XilunWu opened the pull request:
## What problem this PR solves?
#97170 fixed `equal` operator return type (old: Tensor, now: bool) by giving it the correct sharding propagation. This is consistent with the `aten::equal` op. However, the correctness only stays at the local result level:
* `equal` op returns True if the local copy of dtensor A equals to the the local copy of dtensor B

This is not the correct semantic of `equal` which should return True if all local copies of A are equal to the corresponding local copies of B.

## What is this PR?

1. For non-participating ranks, if the return type is scalar, `local_results` is set to `None` which means the default value is a reduced result of participating ranks only.
2. For all ranks, if the return type is scalar and the `op_call` is `aten::equal`(because `aten::equal` is the only function that returns scalar value and needs communication), all gather the `local_results` within the `default pg` and reduce on them with `operator.and_`. The result will be the new `local_result`.

## Result/Impact
For non-participating ranks and the return type is scalar:

1. op is `aten::equal`, the return value is same with all other ranks
2. op is not `aten::equal`, the return value is None. Before this PR, this will raise "NotImplementedError" but has not been tested.

For participating ranks and the return type is scalar:

1. op is `aten::equal`, the return value is the equality of two dtensor operands - True if all copies are equal, False otherwise.
2. op is not `aten::equal`, simply the local computation result.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99014

XilunWu committed:
[BE][DTensor] fix DTensor equal op

[ghstack-poisoned]

XilunWu committed:
Update on "[BE][DTensor] fix DTensor equal op"




[ghstack-poisoned]

XilunWu committed:
Update on "[BE][DTensor] fix DTensor equal op"




[ghstack-poisoned]

XilunWu committed:
Update on "[BE][DTensor] fix DTensor equal op"




[ghstack-poisoned]

wanchaol reviewed (commented):
See comments inlined, can you add some description to the PR's summary about what was wrong before and what the fix is?

XilunWu commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -255,20 +245,26 @@ def default_tensor(spec: DTensorSpec) -> torch.Tensor:
         if op_call in random_ops and is_rng_supported_mesh(mesh):
             set_post_op_offset(dtensor_arg._spec, old_offset)
 
+    # NOTE: can we move this logic into wrap() function???
```
https://github.com/pytorch/pytorch/blob/master/torch/distributed/_tensor/dispatch.py#L36-L73

IIUC this `wrap` function performs post-processing on `local_results` so I think it is reasonable to put the `all-gather` logic there. But after a second thought, I feel it's not very convincing to do that. I'll remove this `NOTE`.

XilunWu committed:
Update on "[BE][DTensor] fix DTensor equal op"




[ghstack-poisoned]

XilunWu commented:
@pytorchbot rebase

XilunWu committed:
Update on "[BE][DTensor] fix DTensor equal op"

## What problem this PR solves?
#97170 fixed `equal` operator return type (old: Tensor, now: bool) by giving it the correct sharding propagation. This is consistent with the `aten::equal` op. However, the correctness only stays at the local result level:
* `equal` op returns True if the local copy of dtensor A equals to the the local copy of dtensor B

This is not the correct semantic of `equal` which should return True if all local copies of A are equal to the corresponding local copies of B.

## What is this PR?

1. For non-participating ranks, if the return type is scalar, `local_results` is set to `None` which means the default value is a reduced result of participating ranks only.
2. For all ranks, if the return type is scalar and the `op_call` is `aten::equal`(because `aten::equal` is the only function that returns scalar value and needs communication), all gather the `local_results` within the `default pg` and reduce on them with `operator.and_`. The result will be the new `local_result`.

## Result/Impact
For non-participating ranks and the return type is scalar:

1. op is `aten::equal`, the return value is same with all other ranks
2. op is not `aten::equal`, the return value is None. Before this PR, this will raise "NotImplementedError" but has not been tested.

For participating ranks and the return type is scalar:

1. op is `aten::equal`, the return value is the equality of two dtensor operands - True if all copies are equal, False otherwise.
2. op is not `aten::equal`, simply the local computation result.




[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -163,9 +164,10 @@ def _operator_dispatch(
-        # from participating devices, and reduce on the list of results
-        # with appropriate operators.
+        # 1. if the return type is scalar, set the local result to None.
+        # The local results from all devices will then be all-gathered
```
nit: can you indent the comments(this line and the below two line) so that it aligned with other comments?

wanchaol reviewed (approved):
lgtm, thanks for the fix! have one minor comment about lint

XilunWu committed:
Update on "[BE][DTensor] fix DTensor equal op"

## What problem this PR solves?
#97170 fixed `equal` operator return type (old: Tensor, now: bool) by giving it the correct sharding propagation. This is consistent with the `aten::equal` op. However, the correctness only stays at the local result level:
* `equal` op returns True if the local copy of dtensor A equals to the the local copy of dtensor B

This is not the correct semantic of `equal` which should return True if all local copies of A are equal to the corresponding local copies of B.

## What is this PR?

1. For non-participating ranks, if the return type is scalar, `local_results` is set to `None` which means the default value is a reduced result of participating ranks only.
2. For all ranks, if the return type is scalar and the `op_call` is `aten::equal`(because `aten::equal` is the only function that returns scalar value and needs communication), all gather the `local_results` within the `default pg` and reduce on them with `operator.and_`. The result will be the new `local_result`.

## Result/Impact
For non-participating ranks and the return type is scalar:

1. op is `aten::equal`, the return value is same with all other ranks
2. op is not `aten::equal`, the return value is None. Before this PR, this will raise "NotImplementedError" but has not been tested.

For participating ranks and the return type is scalar:

1. op is `aten::equal`, the return value is the equality of two dtensor operands - True if all copies are equal, False otherwise.
2. op is not `aten::equal`, simply the local computation result.




[ghstack-poisoned]

XilunWu committed:
Update on "[BE][DTensor] fix DTensor equal op"

## What problem this PR solves?
#97170 fixed `equal` operator return type (old: Tensor, now: bool) by giving it the correct sharding propagation. This is consistent with the `aten::equal` op. However, the correctness only stays at the local result level:
* `equal` op returns True if the local copy of dtensor A equals to the the local copy of dtensor B

This is not the correct semantic of `equal` which should return True if all local copies of A are equal to the corresponding local copies of B.

## What is this PR?

1. For non-participating ranks, if the return type is scalar, `local_results` is set to `None` which means the default value is a reduced result of participating ranks only.
2. For all ranks, if the return type is scalar and the `op_call` is `aten::equal`(because `aten::equal` is the only function that returns scalar value and needs communication), all gather the `local_results` within the `default pg` and reduce on them with `operator.and_`. The result will be the new `local_result`.

## Result/Impact
For non-participating ranks and the return type is scalar:

1. op is `aten::equal`, the return value is same with all other ranks
2. op is not `aten::equal`, the return value is None. Before this PR, this will raise "NotImplementedError" but has not been tested.

For participating ranks and the return type is scalar:

1. op is `aten::equal`, the return value is the equality of two dtensor operands - True if all copies are equal, False otherwise.
2. op is not `aten::equal`, simply the local computation result.




[ghstack-poisoned]

XilunWu commented:
@pytorchmergebot merge -f "CI failure not related to PR"
